ssm_cache/cache.py

```diff
@@ -5,16 +5,101 @@
 from functools import wraps
 import six
 
-import boto3
-
 class InvalidParameterError(Exception):
     """ Raised when something's wrong with the provided param name """
 
+class SSMParameterStore(object): # pylint: disable=too-few-public-methods
+    """Some hooks are provided to override behavior: see inject_session_factory and inject_client_factory
+    for details."""
+
+    @classmethod
+    def _default_session_factory(cls):
+        """Default session factory that creates a boto3 session."""
+        import boto3
+        return boto3.session.Session()
+
+    @classmethod
+    def _default_client_factory(cls, session, name):
+        """Default client factory that creates a client from the provided session."""
+        return session.client(name)
+
+    _SESSION_FACTORY = _default_session_factory
```

Comment:
Shouldn't `_SESSION_FACTORY` be `None` by default, given the implementation of `_session`? It looks like `_default_session_factory` is used when `_SESSION_FACTORY` is `None`, so I would simply set `_SESSION_FACTORY = None` here. Which would allow us to understand whether a custom factory is being used or not.

Alternatively, we can just simplify the implementation of `_session`. Something like:

```python
def _session(cls):
        if cls._SESSION is None:
            if not cls._SESSION_FACTORY:
                raise SomeSpecialException("Missing session factory")
            cls._SESSION = cls._SESSION_FACTORY()
        return cls._SESSION
```

Reply:
I'd probably go with the second option (i.e. keeping the current value and simplifying `_session` with explicit failure instead of silent fallback).

Reply:
@glb changed it from _SESSION_FACTORY = None because pylint complains. With the injector methods, I don't think you even need to check if _SESSION_FACTORY is none.

Reply:
Agreed, let's make pylint happy :)

Reply:
I've simplified `_client` and `_session` to remove the fallback when the factory is `None`, as that shouldn't be possible after adding validation to the injector, barring people setting the factories manually without the injector.